File: CS/CustomFunctionCriteriaOperator.Module/DatabaseUpdate/Updater.cs
using System;

using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Updating;
using DevExpress.Xpo;
using DevExpress.Data.Filtering;
using DevExpress.Persistent.BaseImpl;
using DevExpress.ExpressApp.Security;

using CustomFunctionCriteriaOperator.Module.BusinessObjects;
using DevExpress.ExpressApp.Security.Strategy;
using DevExpress.ExpressApp.Security.Strategy.PermissionMatrix;

namespace CustomFunctionCriteriaOperator.Module.DatabaseUpdate {
   public class Updater : ModuleUpdater {
      public Updater(IObjectSpace objectSpace, Version currentDBVersion)
         : base(objectSpace, currentDBVersion) {
      }
      public override void UpdateDatabaseAfterUpdateSchema() {
         base.UpdateDatabaseAfterUpdateSchema();

         if (ObjectSpace.CreateCollection(typeof(Company)).Count == 0) {

             SecuritySystemRole adminRole = ObjectSpace.FindObject<SecuritySystemRole>(new BinaryOperator("Name", SecurityStrategy.AdministratorRoleName));
            if (adminRole == null) {
                adminRole = ObjectSpace.CreateObject<SecuritySystemRole>();
                adminRole.Name = SecurityStrategy.AdministratorRoleName;
                adminRole.IsAdministrative = true;
            }
            SecuritySystemRole userRole = ObjectSpace.FindObject<SecuritySystemRole>(new BinaryOperator("Name", "User"));
            if (userRole == null) {
                userRole = ObjectSpace.CreateObject<SecuritySystemRole>();
                userRole.Name = "User";
                SecuritySystemTypePermissionObject userTypePermission =
            ObjectSpace.CreateObject<SecuritySystemTypePermissionObject>();
                userTypePermission.TargetType = typeof(SecuritySystemUser);                
                SecuritySystemObjectPermissionsObject currentUserObjectPermission =
                    ObjectSpace.CreateObject<SecuritySystemObjectPermissionsObject>();
                currentUserObjectPermission.Criteria = "[Oid] = CurrentUserId()";
                currentUserObjectPermission.AllowNavigate = true;
                currentUserObjectPermission.AllowRead = true;
                userTypePermission.ObjectPermissions.Add(currentUserObjectPermission);
                userRole.TypePermissions.Add(userTypePermission);
                
                SecuritySystemTypePermissionObject validatedObjectTypePermission =
ObjectSpace.CreateObject<SecuritySystemTypePermissionObject>();
                validatedObjectTypePermission.TargetType = typeof(ValidatedObject);
                validatedObjectTypePermission.AllowWrite = true;
                validatedObjectTypePermission.AllowNavigate = true;
                validatedObjectTypePermission.AllowCreate = true;
                validatedObjectTypePermission.AllowDelete = true;
                validatedObjectTypePermission.AllowRead = true;
                
                userRole.TypePermissions.Add(validatedObjectTypePermission);

                SecuritySystemTypePermissionObject companyTypePermission =
ObjectSpace.CreateObject<SecuritySystemTypePermissionObject>();
                companyTypePermission.TargetType = typeof(Company);
                companyTypePermission.AllowRead = true;

                userRole.TypePermissions.Add(companyTypePermission);
            }

            Employee admin = ObjectSpace.CreateObject<Employee>();
            admin.UserName = "Administrator";
            admin.SetPassword("");
            admin.Roles.Add(adminRole);
            admin.Save();

            Company company1 = ObjectSpace.CreateObject<Company>();
            company1.Name = "Company 1";
            company1.Employees.Add(admin);
            company1.Save();

            Employee user = ObjectSpace.CreateObject<Employee>();
            user.UserName = "Sam";
            user.SetPassword("");
            user.Roles.Add(userRole);
            user.Save();

            Employee user2 = ObjectSpace.CreateObject<Employee>();
            user2.UserName = "John";
            user2.SetPassword("");
            user2.Roles.Add(userRole);
            user2.Save();

            Company company2 = ObjectSpace.CreateObject<Company>();
            company2.Name = "Company 2";
            company2.Employees.Add(user);
            company2.Employees.Add(user2);
            company2.Save();
         }

      }
   }
}

File: VB/CustomFunctionCriteriaOperator.Module/DatabaseUpdate/Updater.vb
Imports System

Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Updating
Imports DevExpress.Xpo
Imports DevExpress.Data.Filtering
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.ExpressApp.Security

Imports CustomFunctionCriteriaOperator.Module.BusinessObjects
Imports DevExpress.ExpressApp.Security.Strategy
Imports DevExpress.ExpressApp.Security.Strategy.PermissionMatrix

Namespace CustomFunctionCriteriaOperator.Module.DatabaseUpdate
   Public Class Updater
       Inherits ModuleUpdater

      Public Sub New(ByVal objectSpace As IObjectSpace, ByVal currentDBVersion As Version)
          MyBase.New(objectSpace, currentDBVersion)
      End Sub
      Public Overrides Sub UpdateDatabaseAfterUpdateSchema()
         MyBase.UpdateDatabaseAfterUpdateSchema()

         If ObjectSpace.CreateCollection(GetType(Company)).Count = 0 Then

             Dim adminRole As SecuritySystemRole = ObjectSpace.FindObject(Of SecuritySystemRole)(New BinaryOperator("Name", SecurityStrategy.AdministratorRoleName))
            If adminRole Is Nothing Then
                adminRole = ObjectSpace.CreateObject(Of SecuritySystemRole)()
                adminRole.Name = SecurityStrategy.AdministratorRoleName
                adminRole.IsAdministrative = True
            End If
            Dim userRole As SecuritySystemRole = ObjectSpace.FindObject(Of SecuritySystemRole)(New BinaryOperator("Name", "User"))
            If userRole Is Nothing Then
                userRole = ObjectSpace.CreateObject(Of SecuritySystemRole)()
                userRole.Name = "User"
                Dim userTypePermission As SecuritySystemTypePermissionObject = ObjectSpace.CreateObject(Of SecuritySystemTypePermissionObject)()
                userTypePermission.TargetType = GetType(SecuritySystemUser)
                Dim currentUserObjectPermission As SecuritySystemObjectPermissionsObject = ObjectSpace.CreateObject(Of SecuritySystemObjectPermissionsObject)()
                currentUserObjectPermission.Criteria = "[Oid] = CurrentUserId()"
                currentUserObjectPermission.AllowNavigate = True
                currentUserObjectPermission.AllowRead = True
                userTypePermission.ObjectPermissions.Add(currentUserObjectPermission)
                userRole.TypePermissions.Add(userTypePermission)

                Dim validatedObjectTypePermission As SecuritySystemTypePermissionObject = ObjectSpace.CreateObject(Of SecuritySystemTypePermissionObject)()
                validatedObjectTypePermission.TargetType = GetType(ValidatedObject)
                validatedObjectTypePermission.AllowWrite = True
                validatedObjectTypePermission.AllowNavigate = True
                validatedObjectTypePermission.AllowCreate = True
                validatedObjectTypePermission.AllowDelete = True
                validatedObjectTypePermission.AllowRead = True

                userRole.TypePermissions.Add(validatedObjectTypePermission)

                Dim companyTypePermission As SecuritySystemTypePermissionObject = ObjectSpace.CreateObject(Of SecuritySystemTypePermissionObject)()
                companyTypePermission.TargetType = GetType(Company)
                companyTypePermission.AllowRead = True

                userRole.TypePermissions.Add(companyTypePermission)
            End If

            Dim admin As Employee = ObjectSpace.CreateObject(Of Employee)()
            admin.UserName = "Administrator"
            admin.SetPassword("")
            admin.Roles.Add(adminRole)
            admin.Save()

            Dim company1 As Company = ObjectSpace.CreateObject(Of Company)()
            company1.Name = "Company 1"
            company1.Employees.Add(admin)
            company1.Save()

            Dim user As Employee = ObjectSpace.CreateObject(Of Employee)()
            user.UserName = "Sam"
            user.SetPassword("")
            user.Roles.Add(userRole)
            user.Save()

            Dim user2 As Employee = ObjectSpace.CreateObject(Of Employee)()
            user2.UserName = "John"
            user2.SetPassword("")
            user2.Roles.Add(userRole)
            user2.Save()

            Dim company2 As Company = ObjectSpace.CreateObject(Of Company)()
            company2.Name = "Company 2"
            company2.Employees.Add(user)
            company2.Employees.Add(user2)
            company2.Save()
         End If

      End Sub
   End Class
End Namespace
